More LabIcon docs

telamonian opened the pull request:
## References

This PR adds a background section, and some sections on creating custom LabIcon to the LabIcon docs in the @jupyterlab/ui-components README.

Since these are narrative docs, and since the jlab API docs are somewhat invisible, these LabIcon docs should probably also go somewhere in the main jlab docs. Where should I put them?

## Code changes

NA

## User-facing changes

NA

## Backwards-incompatible changes

NA

jtpio commented:
Thanks!

> Since these are narrative docs, and since the jlab API docs are somewhat invisible, these LabIcon docs should probably also go somewhere in the main jlab docs. Where should I put them?

At the moment there seems to be two options:

- as a separate page under "Developer Guide", with a link from the "Common Extension Points" page
- inlined in the https://jupyterlab.readthedocs.io/en/stable/developer/extension_points.html#icons section

telamonian commented:
In b921254 I came up with a way to share the same `labicon.md` source file between the API docs and the main jlab docs that I'm reasonably happy with.

Problem is, now the API docs are a bit broken: the markdown link to `labicon.md` that I added to `ui-components/README.md` only works locally, it doesn't work in the rendered API docs.

Gotta think of something else

telamonian commented:
Alright, I think that's a wrap for this PR. Here's the user facing changes (all to docs):

- developer docs
  - added several sections to the icons docs
    - a couple of background sections
    - a couple of sections about creating your own `LabIcon` instances
  - I broke out the LabIcon docs, along with any future ui-component docs, out into their own "Reusing JupyterLab UI" section
  - The previously existing "Icons" section in "Common Extension Points" is now just a link to the above

- API docs
  - as per the dev docs, added the same set of new sections to the LabIcon docs

In terms of non-visible changes, I didn't want to have to keep manually syncing up two nearly identical LabIcon sections in the dev and API docs, so I came up with various ways to generate both from a single source. This turned out to be fairly tricky (one is markdown, the other is rst, they both needed different document titles, etc). I don't think I've yet found the perfect way of doing this, but what I currently have is good enough for this PR.

Briefly, there's now a `docs:init` command in `@jupyterlab/ui-components`. Calling this will rebuild both `packages/ui-components/README.md` and `docs/source/developer/ui_components.rst` from a set of document sources that can be found at `packages/ui-components/docs`

telamonian commented:
One possibly much simpler alternative would be to link to the API docs from the dev docs, or vice versa. I'll look into this and other possibilities while I work on making the API docs nicer overall over at jupyterlab/jupyterlab#8239

saulshanabrook commented on docs/source/developer/ui_components.rst:
```diff
@@ -0,0 +1,337 @@
+     svgstr: fooSvgstr
+   });
+
+Sync icon color to JupyterLab theme
```
I am not too keen about having duplicate text in three places checked into VCS, because it seems likely we might update it in one place and forget the other two. Any ideas on how to make this have a single source of truth?

telamonian replied:
> Any ideas on how to make this have a single source of truth?

Nope, but I'll take suggestions :)

More seriously, I do have some notions about how to improve the `docs:init` build here, but they're all waaay out of scope for what's supposed to be a PR updating the LabIcon docs. I am actively working on improving the API docs, and may be able to fold this issue into that work as well.

For now, we treat the sources in `ui-components/docs` as the source of truth. Running `jlpm docs:init` (while in `ui-components`) will automatically propagate any changes to the README and the dev docs. If anyone does end up commiting changes directly to the wrong copy, I'll deal with manually resolving the conflicts

telamonian replied:
I've now added a warning comment to the top of the autogenerated files:

```html
<!--
THIS FILE IS AUTOGENERATED, DO NOT EDIT

Instead, make changes to docs sources in `packages/ui-components/docs`,
then run "jlpm docs:init" to refresh the built docs
-->
```

saulshanabrook replied:
> Running jlpm docs:init (while in ui-components) will automatically propagate any changes to the README and the dev docs. If anyone does end up commiting changes directly to the wrong copy, I'll deal with manually resolving the conflicts

For the README maybe we could just link to that file instead of copying? And maybe for the docs we could have it do that live generation when generating the docs instead of storing it in VCS?

What do you think?

telamonian replied:
> For the README maybe we could just link to that file instead of copying? And maybe for the docs we could have it do that live generation when generating the docs instead of storing it in VCS?

As I said in an earlier comment, that is exactly how I had things set up earlier in this PR at b921254. The problem was that linking from `ui-components/README.md` to another markdown file doesn't work in the rendered API docs.

In the future I may end up just adding a link in the README.md to the live rendered LabIcon dev docs on readthedocs. However, that will be completely broken until the new LabIcon docs have been pushed to readthedocs in the first place, so I think it's best to do that in the next PR

telamonian replied:
basically, I don't want to roll something project wide and complicated that I'm just going to remove in a future PR. Keeping the docs checked in in 3 places is dumb, but it's also simple and renders reliably

saulshanabrook commented on docs/source/developer/ui_components.rst:
```diff
@@ -0,0 +1,337 @@
+.. raw:: html
+
+   <ul>
+   <li>jp-icon-accent0: <svg width="16" viewBox="0 0 1 1"><rect width="1" height="1" fill="#fff"/><title>#fff</title></svg> / <svg width="16" viewBox="0 0 1 1"><rect width="1" height="1" fill="#111"/><title>#111</title></svg></li>
```
Not sure if we should hard code the color values here. Maybe instead say semantically what the accents are for? Or link to where they are defined? Just imagining if we ever change these values we might forget to update the docs?

telamonian replied:
This is supposed to render as a list of color swatches. It shows up correctly in the built dev docs/my local markdown preview:

![image](https://user-images.githubusercontent.com/2263641/80433012-c1d23300-88c3-11ea-89e3-f89d66b51f1b.png)

but it doesn't look like it's showing up correctly on github (the inline svgs are getting scrubbed). What's it look like to you?

telamonian replied:
I'm of the opinion that a picture is worth a thousand semantic explanations, especially when describing anything visual/having to do with design. So I think the potential tradeoffs are worth it... assuming the swatches render correctly.

These particular colors haven't changed in the last couple of years, and aren't likely to change significantly/frequently in the future. Longer term, we can look into auto-generating the documentation color swatches from the theme CSS.

saulshanabrook replied:
> but it doesn't look like it's showing up correctly on github (the inline svgs are getting scrubbed). What's it look like to you?

Yeah it didn't look like colors to me.

> These particular colors haven't changed in the last couple of years, and aren't likely to change significantly/frequently in the future. Longer term, we can look into auto-generating the documentation color swatches from the theme CSS.

OK sounds good to me! I am convinced.

saulshanabrook commented on docs/source/developer/ui_components.rst:
```diff
@@ -0,0 +1,337 @@
+   <li>jp-icon-accent4: <svg width="16" viewBox="0 0 1 1"><rect width="1" height="1" fill="#757575"/><title>#757575</title></svg> / <svg width="16" viewBox="0 0 1 1"><rect width="1" height="1" fill="#757575"/><title>#757575</title></svg></li>
+   </ul>
+
+For light/dark themes, ``jp-icon-accent0`` corresponds to the
+lighest/darkest background color, while ``jp-icon-accent1`` is somewhat
+darker/lighter, and so forth.
```
Yeah I feel like this is enough!

telamonian replied:
Yep, I think the swatches get the point across more clearly here anyway

saulshanabrook commented on docs/source/developer/ui_components.rst:
```diff
@@ -0,0 +1,337 @@
+Icon handling in Jupyterlab
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Pre jlab-2.0, most icons were created using the icons-as-css-background
```
```suggestion
Pre JupyterLab version 2.0, most icons were created using the icons-as-css-background
```

saulshanabrook commented on docs/source/developer/ui_components.rst:
```diff
@@ -0,0 +1,337 @@
+icon system introduced in JupyterLab v2.0.
+
+Background
+~~~~~~~~~~
```
Could we move the background to the end? I really appreciate it, but think it's probably best to just show ppl how to use it first and foremost, and then they can read about the history after.

telamonian replied:
sure

saulshanabrook commented on docs/source/developer/ui_components.rst:
```diff
@@ -0,0 +1,337 @@
+   import { jupyterIcon } from '@jupyterlab/ui-components';
+
+How to render an icon into a DOM node
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Icons can be added as children to any ``div`` or ``span`` nodes using
+the ``icon.element(...)`` method (where ``icon`` is any instance of
+``LabIcon``). For example, to render the Jupyter icon you could do:
+
+.. code:: typescript
+
+   jupyterIcon.element({
+     container: elem,
+     height: '16px',
+     width: '16px',
+     marginLeft: '2px'
+   });
+
+where ``elem`` is any ``HTMLElement`` with a ``div`` or ``span`` tag. As
+shown in the above example, the icon can be styled by passing CSS
+parameters into ``.element(...)``. Any valid CSS parameter can be used
+(one catch: snake case params do have to be converted to camel case:
+instead of ``foo-bar: '8px'``, you’d need to use ``fooBar: '8px'``.
+
+How to render an icon as a React component
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
I felt like there were more ways to use it that these two, but maybe I was mis-remembering? Like is there some way we use it by the "name"? Or what function does the name play?

telamonian replied:
> Like is there some way we use it by the "name"?

Yes, there are the `LabIcon.resolveElement` and `LabIcon.resolveReact` static methods. The use case for these is very narrow; so far the only use case I've actually seen out in extensionland is for reusing our FileType icons in custom browser implementations. There's also a bunch of ways to shoot yourself in the foot with these. For now, I'm comfortable leaving the `resolveX` methods in the [API docs](http://jupyterlab.github.io/jupyterlab/ui-components/classes/labicon.html#resolveelement), where they are well covered...

at least that's what I would have liked to have said. I just now took a closer look at the API docs, and it looks like most of my docstrings are getting butchered by typedoc. Whelp, guess that's another thing to put on the list of "shit to fix in the API docs"

saulshanabrook replied:
OK! So are all the other methods/attributes (like `svgstr`) effectively deprecated?

telamonian replied:
? I don't quite follow your logic there in declaring `svgstr` deprecated. There are a handful of public fields (`name`, `svgstr`, `render`) in a LabIcon instance that exist for the purpose of exposing metadata and/or integrating with external projects, eg `render` is used in the Lumino integration. `name` and `svgstr` are also exposed for the purpose of consistency across icon interfaces, so that eg `IFileType.icon` in `rendermime-interfaces` can take either a `LabIcon` instance or a simple `{name, svgstr}` pair. They aren't relevant for the extension dev docs, though I may add some more info about them to the API docs.

telamonian commented:
I want the improved LabIcon docs to be available to people ASAP, so I'm going to pull this in now and save the improvements to how those docs are built for a later PR

blink1073 commented:
@meeseeksdev please backport to 2.1.x
